# bpertussisciwgs: Output

## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

Pipeline outputs are organized within the user-defined `--outdir` and all paths are relative to this top-level results directory.
```
{outdir}/
├── fastqc/
├── filteredreads/
│   └── {sample}.filtered.R[12]_001.fastq.gz     -- Filtered B. pertussis reads for downstream analyses
├── mapping-bp/
│   ├── {sample}.bp-depth.tsv                    -- Coverage depth across the B. pertussis reference genome (samtools depth)
│   ├── {sample}.bp-mapped.sorted.bam            -- Read mapping to the B. pertussis reference genome
│   └── {sample}.bp-mapped.sorted.bam.bai
├── mapping-hs/
│   ├── {sample}.hs-unmapped.bam.                -- Filtered bam of reads NOT mapping to human reference
│   ├── {sample}.hs-unmapped.R[12].fastq.gz      -- Filtered reads NOT mapping to human reference
│   └── {sample}.hs-unmapped.flagstat
├── quast/
├── multiqc/
├── pipeline_info/
├── quast/
│   └── report.tsv                               -- Assembly quality metrics from QUAST
├── skesa/
│   ├── {sample}.skesa.fasta                     -- De novo assembly from filtered B. pertussis reads using skesa
│   └── {sample}.skesa.wgMLST.bls                -- BLASTn alignment of wgMLST reference alleles to skesa assembly
│
└── sample_summary.tsv                           -- Summary table of quality metrics (see below)

```


## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

* [FastQC](#fastqc) - Raw read QC
* [mapping-hs](#mapping-hs) - Read mapping
* [mapping-bp](#mapping-bp) - Read filtering
* [skesa](#skesa) - De novo assembly
* [mlst](#mlst) - MLST
* [MultiQC](#multiqc) - Aggregate report describing results and QC from the whole pipeline
* [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution
* [Sample summary](#sample-summary) - Sample summary of quality metrics

### FastQC

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).  
<details markdown="1">
<summary>Output files</summary>

* `fastqc/`
    * `*_fastqc.html`: FastQC report containing quality metrics.
    * `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

</details>


### Mapping-HS
<details markdown="1">
<summary>Output files</summary>

* `mapping-hs/`
    * `*.hs-unmapped.bam`: 
    * `*.hs-unmapped.flagstat`: 
    * `*.hs-unmapped.R[12].fastq.gz`:
    * `*.bowtie2.log`:

</details>

### Mapping-BP
<details markdown="1">
<summary>Output files</summary>

* `mapping-bp/`
    * `*.bp-mapped.[sorted].bam`: 
    * `*.bp-mapped.flagstat`: 
    * `*.bp-depth.tsv`:
    * `*.bowtie2.log`:
    * `reference/*bt2`:
* `filteredreads/`
    * `*.filtered.R[12].fastq.gz`:

</details>

### MLST
<details markdown="1">
<summary>Output files</summary>

* `mlst/`
    * `[reference].mlst.bed`:
    * `[reference].mlst.bls`:
    * `*.mlst-bedcov.tsv`: 

</details>

### Skesa
<details markdown="1">
<summary>Output files</summary>

* `skesa/`
    * `*.skesa.fasta`: 
    * `*.skesa.log`: 

</details>

### QUAST
<details markdown="1">
<summary>Output files</summary>

* `quast/`
    * `quast/`:
    * `transposed_report.tsv`:

</details>

### MultiQC

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.
<details markdown="1">
<summary>Output files</summary>

* `multiqc/`
    * `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
    * `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
    * `multiqc_plots/`: directory containing static images from the report in various formats.

</details>


### Pipeline information

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
<details markdown="1">
<summary>Output files</summary>

* `pipeline_info/`
    * Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.svg`/`pipeline_dag.svg`.
    * Reports generated by the pipeline: `pipeline_report.html` and `software_versions.yml`.
    * Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.

</details>


### Sample Summary
<details markdown="1">
<summary>Output files</summary>

* `./sample_summary.tsv`: Sample summary of quality metrics

|Field|Description|
|-----|-----|
|Sample|Sample name|
|Reads-total|Total count of input read pairs|
|Reads-Hs|Count of read pairs mapped to human reference|
|Frac-Hs|Fraction of read pairs mapped to human reference|
|Reads-Bp|Count of read pairs mapped to *B. pertussis* reference|
|Frac-Bp|Fraction of read pairs mapped to *B. pertussis* reference|
|Reads-Unmap|Count of read pairs not mapped to either reference|
|Frac-Unmap|Fraction of read pairs not mapped|
|Depth|Average *B. pertussis* sequencing depth|
|Breadth-20x-len|Number of positions in *B. pertussis* reference with >=20x coverage|
|Breadth-20x-frac|Fraction of positions in *B. pertussis* reference with >=20x coverage|
|Breadth-50x-len|Number of positions in *B. pertussis* reference with >=50x coverage|
|Breadth-50x-frac|Fraction of positions in *B. pertussis* reference with >=50x coverage|
|MLST-20x|*Optional:* Number of typing loci with >=20x coverage|
|MLST-50x|*Optional:* Number of typing loci with >=50x coverage|
|Read-len|Average raw read length (bp)|
|Asm-contigs|*Optional:* De novo assembly contig count|
|Asm-length|*Optional:* De novo assembly total length (bp)|
|Asm-N50|*Optional:* De novo assembly N50 (bp)|
|Asm-longest|*Optional:* De novo assembly longest contig length (bp)|
|Asm-Ref-frac|*Optional:* De novo assembly aligned fraction of reference|

</details>